convert2rhel/applock.py

```diff
@@ -0,0 +1,178 @@
+    again.
+
+    For safety, unexpected conditions, like garbage in the file or
+    bad permissions, are treated as if the application is locked,
+    because something is obviously wrong.
+    """
+
+    def __init__(self, name):
+        # Our application name
+        self._name = name
+        # Do we think we locked the pid file?
+        self._locked = False
+        # Maximum number of tries to lock
+        self._max_loop_count = 5
+        # Our process ID
+        self._pid = os.getpid()
+        # Path to the file that contains the process id
+        self._pidfile = os.path.join(_DEFAULT_LOCK_DIR, self._name + ".pid")
+
+    def __str__(self):
+        if self._locked:
+            status = "locked"
+        else:
+            status = "unlocked"
+        return "%s PID %d %s" % (self._pidfile, self._pid, status)
+
+    def _try_create(self):
+        """Try to create the lock file. If this succeeds, the lock file
+        exists and we created it. If an exception other than the one
+        we expect is raised, re-raises it.
+
+        :returns: True if we created the lock, False if we didn't.
+        """
+        #
+        # Create a temporary file that contains our PID and attempt
+        # to link it to the real pidfile location. The link will fail if
+        # the file already exists; this avoids a race condition when
+        # two processes attempt to create the file simultaneously. This
+        # also guarantees that the lock file contains valid data.
+        #
+        with tempfile.NamedTemporaryFile(mode="wt", suffix=".pid", prefix=self._name, dir=_DEFAULT_LOCK_DIR) as fileh:
+            fileh.write(str(self._pid) + "\n")
+            fileh.flush()
+            try:
+                os.link(fileh.name, self._pidfile)
+            # In Python 3 this could be changed to catch FileExistsError.
+            except OSError as exc:
+                if exc.errno == errno.EEXIST:
+                    return False
+                raise exc
+        loggerinst.debug("%s." % self)
+        return True
+
+    @property
+    def is_locked(self):
+        """Test whether this object is locked."""
+        return self._locked
+
+    @staticmethod
+    def _pid_exists(pid):
+        """Test whether a particular process exists."""
+        try:
+            os.kill(pid, 0)
+        except OSError:
+            return False
+        return True
+
+    def trylock(self, loop_count=0):
+        """Try to get a lock on this application. If successful,
+        the application will be locked; the lock should be released
+        with unlock().
+
+        If the file has unexpected contents, for safety we treat the
+        application as locked, since it is probably the result of
+        manual meddling, intentional or otherwise.
+
+        Note that nothing prevents you from calling trylock()
+        multiple times; all calls after the first will fail as if
+        another process holds the lock.
```

Comment:
It's an artifact of evolution: I originally added self.locked just for use by the unit tests, then thought to use it so you can call unlock() without holding the lock--required for my original use of atexit under 2.7 where there is no unregister call.

If you want to lock separate sections of the code you can create ApplicationLocks with two different names and they should be completely separate (apart from the variable that controls the lock directory), so the lock state of one will be independent of the other. Though you'd have to worry about ordering for obvious reasons.

Given the one-to-one correspondence between an instance of an ApplicationLock and a single file in the file system, I think something like a reference count would best be served by a completely different design, because it's a different requirement than "keep two instances of this program from running at the same time".